File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private int _battery = 100;
    private int _distance;
    private int _speed;
    private int _batteryDrain;
    
    public RemoteControlCar(int speed, int batteryDrain)
    {
        _batteryDrain = batteryDrain;
        _speed = speed;
    }
    
    public bool BatteryDrained() => (_battery < _batteryDrain);

    public int DistanceDriven() => _distance;

    public void Drive()
    {
        if (!BatteryDrained())
        {
            _battery -= _batteryDrain;
            _distance += _speed;
        }  
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int _distance;
    
    public RaceTrack(int distance)
    {
        _distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained()) {car.Drive();}
        return car.DistanceDriven() >= _distance;
    }
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed 
{

    private int _speed;
    private int _battery = 100;
    private int _batteryDrain;
    private int _distanceDriven;
    
    NeedForSpeed(int speed, int batteryDrain) 
    {
        _speed = speed;
        _batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() 
    {
        return _battery < _batteryDrain;
    }

    public int distanceDriven() 
    {
        return _distanceDriven;
    }

    public void drive() 
    {
        if (!batteryDrained()) 
        { 
            _distanceDriven += _speed; 
            _battery -= _batteryDrain;
        }
    }

    public static NeedForSpeed nitro() 
    {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack 
{

    private int _distance;
    
    RaceTrack(int distance) 
    {
        _distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) 
    {
        while (!car.batteryDrained()) { car.drive(); }
        return car.distanceDriven() >= _distance;
        
    }
}
